File: android/app/src/main/java/io/github/mattlavallee/ratify/core/Group.kt
package io.github.mattlavallee.ratify.core

import com.google.android.gms.location.places.Place
import java.util.Date

class Group {
    var id: String
    var type: String
    var name: String
    var description: String
    var activity: String
    var placeName: String = ""
    var placeLatitude: Double = -1.0
    var placeLongitude: Double = -1.0
    var maxResults: Int = -1
    var expirationDays: Int = -1
    var voteConclusion: Date = Date(0)
    var participants: Int = -1

    constructor(
        id: String,
        type: String,
        name: String,
        descr: String,
        activity: String,
        location: Place?,
        numResults: Int,
        conclusion: Date,
        expiration: Int
    ) {
        this.id = id
        this.type = type
        this.name = name
        this.description = descr
        this.activity = activity
        if (location != null) {
            this.placeName = location.name.toString()
            this.placeLatitude = location.latLng.latitude
            this.placeLongitude = location.latLng.longitude
        }

        this.maxResults = numResults
        this.voteConclusion = conclusion
        this.expirationDays = expiration
    }

    constructor(
        id: String,
        type: String,
        name: String,
        descr: String,
        activity: String,
        latitude: Double,
        longitude: Double,
        numResults: Int,
        conclusion: Date,
        expiration: Int,
        participants: Int?) {
        this.id = id
        this.type = type
        this.name = name
        this.description = descr
        this.activity = activity
        this.placeLatitude = latitude
        this.placeLongitude = longitude
        this.maxResults = numResults
        this.voteConclusion = conclusion
        this.expirationDays = expiration
        if (participants != null) {
            this.participants = participants
        }
    }

    fun isConcluded(): Boolean {
        return this.voteConclusion.before(Date())
    }

    fun populateParams(params: MutableMap<String, Any>): MutableMap<String, Any> {
        params["name"] = this.name
        params["type"] = this.type
        params["description"] = this.description
        params["activity"] = this.activity
        params["startingLocation"] = this.placeName
        params["latitude"] = this.placeLatitude
        params["longitude"] = this.placeLongitude
        params["results"] = this.maxResults
        params["expiration"] = this.expirationDays
        params["conclusion"] = this.voteConclusion.time
        return params
    }

    companion object {
        fun fromJsonHashMap(id: String, model: HashMap<String, Any>): Group {
            @Suppress("UNCHECKED_CAST")
            val location: HashMap<String, Double> = model["location"] as HashMap<String, Double>

            var totalActiveParticipants = 0
            if (model["members"] != null) {
                @Suppress("UNCHECKED_CAST")
                val participants: HashMap<String, Boolean> = model["members"] as HashMap<String, Boolean>

                totalActiveParticipants = participants.count { it.value }
            }

            var query = if (model["query"] != null) model["query"] as String else model["activity"] as String
            var numberResults = if(model["numberResults"] != null) model["numberResults"] as Int else model["maxResults"] as Int
            var expiration = if(model["daysToExpire"] != null) model["daysToExpire"] as Int else model["expiration"] as Int
            var conclusion = if(model["voteConclusionEpoch"] != null) model["voteConclusionEpoch"] as Long else model["voteConclusion"] as Long

            return Group(
                id,
                model["type"] as String,
                model["name"] as String,
                model["description"] as String,
                query,
                location["latitude"] as Double,
                location["longitude"] as Double,
                numberResults,
                Date(conclusion),
                expiration,
                totalActiveParticipants
            )
        }
    }
}

File: firebase/functions/src/models/group.ts
import { IDetailedGroupVotes } from "./vote";
import { IMatchDetails, ISingleMatch } from "./match";

export interface IGroup {
  name: string,
  type: string,
  query: string,
  description: string,
  location: {
    latitude: number,
    longitude: number,
  },
  numberResults: number,
  voteConclusion: number,
  daysToExpire: number,
  members: {
    [userId: string]: boolean
  },
  matches: {
    [matchId: string]: boolean
  }
};

export interface IUserGroup {
  created_groups: {[groupId: string]: IGroup},
  joined_groups: {[groupId: string]: IGroup},
};

export class GroupRequest {
  public name: string;
  public type: string;
  public description: string;
  public activity: string;
  public location: {[key: string]: any} = {
    name: '',
    latitude: 0,
    longitude: 0
  };
  public maxResults: number;
  public voteConclusion: Date;
  public expiration: number;

  constructor(name: string, type: string, descr: string, activity: string, location: string, 
             latitude: number, longitude: number, results: number, conclusion: number, 
             expiration: number) {
    this.name = name;
    this.type = type;
    this.description = descr;
    this.activity = activity;
    this.location.name = location;
    this.location.latitude = latitude;
    this.location.longitude = longitude;
    this.maxResults = results;
    this.voteConclusion = new Date(conclusion);
    this.expiration = expiration;
  }

  isValid() {
    if (this.name && this.name.length &&
        this.type && this.type.length &&
        this.description && this.description.length &&
        this.activity && this.activity.length
        && this.location.name.length &&
        this.maxResults > 0 && this.maxResults <= 30 &&
        !isNaN(this.voteConclusion.getTime()) &&
        this.expiration > 0 && this.expiration <= 14) {
      return true;
    }

    return false;
  }
}

export class DetailedGroup extends GroupRequest {
  public userVotes: IDetailedGroupVotes = {};
  public matches: {[matchId: string]: IMatchDetails} = {};

  constructor(group: IGroup) {
    super(group.name, group.type, group.description, group.query, '_', group.location.latitude,
      group.location.longitude, group.numberResults, group.voteConclusion, group.daysToExpire);
  }

  public setVoteState(votes: IDetailedGroupVotes) {
    this.userVotes = votes;
  }

  public setMatches(inflatedMatches: ISingleMatch[]) {
    inflatedMatches.forEach((match: ISingleMatch) => {
      this.matches[match.details.id] = match.details;
    });
  }
}